internal/p2p/channel.go

```diff
@@ -259,17 +264,85 @@ func (ch *libp2pChannelImpl) canonicalizedTopicName() string {
 }
 
 func (ch *libp2pChannelImpl) Receive(ctx context.Context) *ChannelIterator {
+	// TODO: consider caching an iterator in the channel, or
+	// erroring if this gets called more than once.
+	//
+	// While it's safe to register a handler more than once, we
+	// could get into a dodgy situation where if you call receive
+	// more  than once, the subsequently messages won't be routed
+	// correctly.
 	iter := &ChannelIterator{
 		pipe: make(chan Envelope),
 	}
 
+	ch.host.SetStreamHandler(protocol.ID(ch.canonicalizedTopicName()), func(stream network.Stream) {
+		buf := bufio.NewReader(stream)
+		// TODO: change the max delimited message into
+		// something that's passed in as a parmeter.
+		reader := protoio.NewDelimitedReader(buf, 1023)
+
+		remote := stream.Conn().RemotePeer()
+
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+
+			payload := proto.Clone(ch.chDesc.MessageType)
+
+			// TODO: context might expire while reading
```

Comment:
If we think that's likely to be a blocker (e.g., we might ignore cancellation indefinitely), this is probably a good case to terminate the stream explicitly:

```go
ctx, cancel := context.WithCancel(ctx)
defer cancel()
go func() { <-ctx.Done(); stream.Close() }()
```

(the extra cancel is just to make sure the closer eventually gets to exit)

Close promises that:
```
// * Any buffered data for writing will be flushed.
// * Future reads will fail.
// * Any in-progress reads/writes will be interrupted.
```

I tried to make sure it's safe to close multiple times, but libp2p abuses interfaces even worse than we do. 😞